examples/arkode/C_serial/ark_analytic_partitioned.c

```diff
@@ -0,0 +1,263 @@
+ * Programmer(s): Steven B. Roberts @ LLNL
+ *------------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2024, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ *------------------------------------------------------------------------------
+ * We consider the initial value problem
+ *    y' + lambda*y = y^2, y(0) = 1
+ * proposed in 
+ * 
+ * Estep, D., et al. "An a posteriori–a priori analysis of multiscale operator
+ * splitting." SIAM Journal on Numerical Analysis 46.3 (2008): 1116-1146.
+ *
+ * The parameter lambda is positive, t is in [0, 1], and the exact solution is
+ * 
+ *    y(t) = lambda*y / (y(0) - (y(0) - lambda)*exp(lambda*t))
+ *
+ * This program solves the problem with a splitting or forcing method which can
+ * be specified with the command line syntax
+ * 
+ * ./ark_analytic_partitioned <integrator> <coefficients>
+ *    integrator: either 'splitting' or 'forcing'
+ *    coefficients (splitting only): the SplittingStepCoefficients to load
+ * 
+ * The linear term lambda*y and nonlinear term y^2 are treated as the two
+ * partitions. The former is integrated with a time step 5x smaller than the
+ * outer method, while the later uses a 10x smaller time step. Once solved, the
+ * program prints the error and statistics.
+ *----------------------------------------------------------------------------*/
+
+/* Header files */
+#include <arkode/arkode_arkstep.h>
+#include <arkode/arkode_forcingstep.h>
+#include <arkode/arkode_splittingstep.h>
+#include <nvector/nvector_serial.h>
+#include <string.h>
+
+#if defined(SUNDIALS_EXTENDED_PRECISION)
+#define GSYM "Lg"
+#define ESYM "Le"
+#define FSYM "Lf"
+#else
+#define GSYM "g"
+#define ESYM "e"
+#define FSYM "f"
+#endif
+
+typedef struct
+{
+  sunrealtype lambda;
+} UserData;
+
+/* RHS for f^1(t, y) = -lambda * y */
+static int f_linear(const sunrealtype t, const N_Vector y, N_Vector ydot,
+                    void* const user_data)
+{
+  sunrealtype lambda = ((UserData*)user_data)->lambda;
+  N_VScale(-lambda, y, ydot);
+  return 0;
+}
+
+/* RHS for f^2(t, y) = y^2 */
+static int f_nonlinear(const sunrealtype t, const N_Vector y, N_Vector ydot,
+                       void* const user_data)
+{
+  N_VProd(y, y, ydot);
+  return 0;
+}
+
+/* Compute the exact analytic solution */
+static N_Vector exact_sol(const N_Vector y0, const sunrealtype tf,
+                          const UserData* const user_data)
+{
+  N_Vector sol             = N_VClone(y0);
+  const sunrealtype y0_val = NV_Ith_S(y0, 0);
+  const sunrealtype lambda = user_data->lambda;
+  NV_Ith_S(sol, 0)         = lambda * y0_val /
+                     (y0_val - (y0_val - lambda) * SUNRexp(lambda * tf));
```

Comment:
I thought that in the arkode-sts PR #541, you requested that @maggul's new examples avoid all uses of `NV_Ith_S`?

Reply:
We want to remove the use of NV_Ith in all examples.

Reply:
Hmm, I think that was David who requested that. Anyway, `NV_Ith_S` has been removed.